### DISCUSSION TITLE
Support Nx `workspace.json` as a valid configuration file

### DISCUSSION THREAD

**Discussion by wSedlacek** [cyrilletuzi/vscode-angular-schematics#458] [Ideas]
It would be nice if `workspace.json` (from @nrwl/nx) was used when `angular.json` is not present.

**Comment by cyrilletuzi** [comment]
From the [Nx documentation](https://nx.dev/latest/angular/getting-started/configuration#workspace-json), some properties are not the same in `workspace.json`:
- `targets` > `architect` (currently not used by this extension)
- `builder` > `executor` (currently not used by this extension)
- `schematics` > `generators` (used by this extension, loaded [here](https://github.com/cyrilletuzi/vscode-angular-schematics/blob/547c7d0cc098d021c02349e781a40ca271935397/src/workspace/angular/angular.ts#L112) and [here](https://github.com/cyrilletuzi/vscode-angular-schematics/blob/547c7d0cc098d021c02349e781a40ca271935397/src/workspace/angular/angular.ts#L99))

So it would require further investigations to know what changes exactly (I don't what what is the exact configuration syntax of Nx, if the `schema.json` describing it could be found it would be nice).

If they are just adding new properties and leaving the Angular ones intact, that could be OK. If they are replacing the Angular ones it would be more risky (`angular.json` could jump to a version `2` someday too).

**Comment by wSedlacek** [comment]
From what I have observed the only setting that this extension looks at is the default collection. Are there other cases where angular.json is parsed that these name changes would be meaningful?

**Comment by cyrilletuzi** [comment]
Yes, `schematics` for default options values, I pointed to the files in my previous message.

**Comment by wSedlacek** [comment]
This might be a good reference 
https://www.github.com/nrwl/nx/tree/master/packages%2Ftao%2Fsrc%2Fshared%2Fworkspace.ts

**Comment by cyrilletuzi** [comment]
Thanks for the reference file.

After some investigation, the Angular configuration [supported files](https://github.com/angular/angular-cli/blob/a7b9497b63f8608e4640f68e24558aa6b20f6f7a/packages/angular_devkit/architect_cli/bin/architect.ts#L197) are `angular.json`, `.angular.json`, `workspace.json`, `.workspace.json`.

Which is very surprising to me, because I'm quite sure I checked that less than one year ago (during the v4 full rewrite of the extension in 2020 spring), but it seems allowed since Angular 6 ([change from march 2018](https://github.com/angular/devkit/pull/554/files)).

So I'll add support for the according files.

On the Nx side, what is worrying me is that the `workspace.json` has never been officially documented by the Angular CLI. So is Nx using this unknown option to add their own specific config (which could cause issues if the Angular CLI itself updates to `version: 2` in some future, and for this extension it would mean we need a special handler for Nx projects), or are the new options also official but unknown? It will need more investigation inside the Angular CLI repo.

**Comment by cyrilletuzi** [comment]
Well after more investigation, the good news is I'm not crazy, the bad news is that the Angular CLI code seems very inconsistent:
- [`packages/angular/cli/utilities/config.ts`](https://github.com/angular/angular-cli/blob/a7b9497b63f8608e4640f68e24558aa6b20f6f7a/packages/angular/cli/utilities/config.ts#L50): `angular.json`, `.angular.json`
- [`packages/angular/cli/utilities/project.ts`](https://github.com/angular/angular-cli/blob/a7b9497b63f8608e4640f68e24558aa6b20f6f7a/packages/angular/cli/utilities/project.ts#L15): `angular.json`, `.angular.json`, `angular-cli.json`, `.angular-cli.json`
- [`packages/angular_devkit/architect_cli/bin/architect.ts`](https://github.com/angular/angular-cli/blob/a7b9497b63f8608e4640f68e24558aa6b20f6f7a/packages/angular_devkit/architect_cli/bin/architect.ts#L197): `angular.json`, `.angular.json`, `workspace.json`, `.workspace.json`
- [`packages/angular_devkit/core/src/workspace/core.ts`](https://github.com/angular/angular-cli/blob/a7b9497b63f8608e4640f68e24558aa6b20f6f7a/packages/angular_devkit/core/src/workspace/core.ts#L38): `angular.json`, `.angular.json`

I'll try to ask an Angular team member.

**Comment by cyrilletuzi** [comment]
After more investigation:
- while still referenced in one unused file, `angular-cli.json` and .`angular-cli.json` are no longer supported by Angular CLI,
- `workspace.json` and `.workspace.json` are not supported in a regular Angular CLI project (1. `ng new`, 2. rename `angular.json` to `workspace.json`, 3. `ng g component hello` > doesn't work), it is a Nx specific feature, which works only because Nx is overriding the `ng` command (you can see it in `node_modules/.bin/ng`).
